architecture_records/004-dataloader-v2.md

```diff
@@ -0,0 +1,314 @@
+# Data Loader Design For FMS-HF-Tuning
+
+**Deciders(s)**: Sukriti Sharma (sukriti.sharma4@ibm.com), Will Johnson (Will.Johnson@ibm.com) , Abhishek Maurya (maurya.abhishek@ibm.com), Dushyant Behl (dushyantbehl@in.ibm.com), Ashok Pon Kumar (ashokponkumar@in.ibm.com)
+
+**Date (YYYY-MM-DD)**:  2024-03-06
+
+**Obsoletes ADRs**:  NA
+
+**Modified By ADRs**:  NA
+
+**Relevant Issues**: [1]
+
+- [Summary and Objective](#summary-and-objective)
+  - [Motivation](#motivation)
+  - [User Benefit](#user-benefit)
+- [Decision](#decision)
+  - [Alternatives Considered](#alternatives-considered)
+- [Consequences](#consequences)
+- [Detailed Design](#detailed-design)
+
+## Summary and Objective
+
+The reason for motivating dataloader design for fms-hf-tuning is to have a unified interface which supports many type of data formats, streaming and non streaming data, weight based data mixing and many others. Full list we focus on is below - 
+
+1. Data Loading -
+
+    1. Different formats of data → Arrow, Parquet etc.
+
+    1. Different modalities of data -> Images, Audio etc.
+
+    1. Streaming data.
+
+    1. Data Replay
+
+    1. Resume with different number of GPUs
```

Comment:
how is this going to happen with preprocessing?

Reply:
The implementation of data loader in the `fms-fsdp` supports resuming with different number of GPUs so when we add support for that dataloader we can pass additional args to the dataloader as the gpu number.